components/base/wheeled/wheeled_base.go

```diff
@@ -91,17 +100,186 @@ type wheeledBase struct {
+		base.right = append(base.right, m)
+	}
+
+	for _, msName := range cfg.MovementSensor {
+		ms, err := movementsensor.FromDependencies(deps, msName)
+		if err != nil {
+			return nil, errors.Wrapf(err, "no movement_sensor namesd (%s)", msName)
+		}
+		base.movementSensor = append(base.movementSensor, ms)
+		props, err := ms.Properties(ctx, nil)
+		if props.OrientationSupported && err == nil {
+			base.orienationSupported = props.OrientationSupported
+			base.orientationSensor = ms
+		}
+	}
+
+	base.allMotors = append(base.allMotors, base.left...)
+	base.allMotors = append(base.allMotors, base.right...)
+
+	var err error
+	base.model, err = base.createModelFrame(baseName, cfg.WidthMM)
+	if err != nil {
+		logger.Error("error creating base model, check that a name for the base and width are provided in the configuration")
+	}
+
+	return base, nil
+}
+
+func (base *wheeledBase) createModelFrame(baseName string, widthMM int) (referenceframe.Model, error) {
+	basePose := spatialmath.NewPose(
+		r3.Vector{X: 0, Y: 0, Z: 0},
+		&spatialmath.OrientationVectorDegrees{Theta: 0, OX: 0, OY: 1, OZ: -90})
+	geometry, err := spatialmath.NewBox(
+		basePose,
+		r3.Vector{X: float64(widthMM), Y: float64(widthMM), Z: float64(widthMM)},
+		"base_dimensions")
+	if err != nil {
+		return nil, err
+	}
+	model := referenceframe.NewSimpleModel(baseName)
+	f, err := referenceframe.NewStaticFrameWithGeometry(baseName, basePose, geometry)
+	if err != nil {
+		return nil, err
+	}
+	model.OrdTransforms = append(model.OrdTransforms, f)
+	return model, nil
+}
+
+func (base *wheeledBase) Spin(
+	ctx context.Context, angleDeg, degsPerSec float64, extra map[string]interface{},
+) error {
+	switch {
+	case base.orientationSensor != nil && base.orienationSupported:
+		base.logger.Debug("spinning with movement sensor")
+		return base.spinWithMovementSensor(ctx, angleDeg, degsPerSec, extra)
+
+	default:
+		base.logger.Debug("spinning without movement sensors")
+		return base.spinWithoutMovementSensor(ctx, angleDeg, degsPerSec)
+	}
+}
+
+func getCurrentYaw(ctx context.Context, ms movementsensor.MovementSensor, extra map[string]interface{},
+) (float64, error) {
+	orientation, err := ms.Orientation(ctx, extra)
+	if err != nil {
+		return 0, err
+	}
+	return rdkutils.RadToDeg(orientation.EulerAngles().Yaw), nil
+}
+
+func (base *wheeledBase) spinWithMovementSensor(ctx context.Context, angleDeg, degsPerSec float64, extra map[string]interface{}) error {
+	revs, wheelrpm := base.spinMath(angleDeg, degsPerSec)
+
+	startYaw, err := getCurrentYaw(ctx, base.orientationSensor, extra)
+	if err != nil {
+		base.logger.Errorf("error: %#v", err)
+		return err
+	}
+	targetYaw := startYaw + angleDeg
+	currYaw := startYaw
+	base.activeBackgroundWorkers.Add(1)
+	utils.ManagedGo(func() {
+		for {
+			select {
+			case <-ctx.Done():
+				base.logger.Debugf("second cancelled context hit %#v", ctx.Err())
+				return
+			default:
+			}
+			errAngle := targetYaw - currYaw
```

Comment:
[rand-note] after testing with hardware for new wrapping angle logic, remove line 249.